### ISSUE TITLE
scaffold.pl: Issue with empty answers in MultiAnswer inside scaffold

### ISSUE THREAD

**Issue by taniwallach** [openwebwork/pg#585]
Due to a student complaint, I discovered that when the first input box of a MultiAnswer group (which has `allowBlankAnswers => 1`) is left empty, then the behavior inside a scaffold section differs from that without the use of scaffolding.

**Update:** seen also with `allowBlankAnswers => 0`, and for all but the last box of the MultiAnswer group.

A sample pair of minimal problems follow below. When the left box is left empty in the scaffolded version, the warning
```
Use of uninitialized value in string ne at line 380 of [PG]/macros/parserMultiAnswer.pl
```
is issued. No such warning is issued on the non-scaffolded version.

**Update: As pointed out by @drgrice1 below, the error was actually in line 378, so lines 379-380 copied below are not relevant to the issue.**

The error message was from testing on a server running a relatively recent copy of `PG-2.16`, where lines 379-380 of `macros/parserMultiAnswer.pl` have:
```
    return if $self->{checkTypes} && $ans->{student_value}->type ne $ans->{correct_value}->type &&
              !($self->{allowBlankAnswers} && $ans->{student_ans} !~ m/\S/) ;
```

---

```
DOCUMENT();

loadMacros(
  "PGstandard.pl",
  "MathObjects.pl",
  "parserMultiAnswer.pl",
  "scaffold.pl",
  "PGcourse.pl",
);

# Print problem number and point value (weight) for the problem
TEXT(beginproblem());

# Show which answers are correct and which ones are incorrect
$showPartialCorrectAnswers = 1;

Context("Numeric");

$a = 2;
$b = 3;

$ans1 = Real($a);
$ans2 = Real($b);

# ===============================================================


$mans1 = MultiAnswer( $ans1, $ans2 )->with(
  singleResult => 0,
  allowBlankAnswers => 1,
  checker => sub {
      my ( $correct, $student, $self, $ansHash ) = @_;
      my @stud_ans = @{$student};
      my @results = ( 0,0 );

      for ( $ii = 0; $ii < 2 ; $ii++ ) {
        if ( $$correct[ $ii ] == $$student[ $ii ] ) {
          $results[ $ii ] = 1;
        }
      }
      return( @results );
  }
);



# ===============================================================


###########################################
#  The scaffold
Scaffold::Begin(
      can_open => "when_previous_correct",
      is_open  => "correct_or_first_incorrect",
      instructor_can_open => "always",
      after_AnswerDate_can_open => "always",
      hardcopy_is_open  => "correct_or_first_incorrect",
    );


###########################################
Section::Begin("Part 1:");

Context()->texStrings;
BEGIN_TEXT

Text removed. There had been a question here. Not needed to trigger the issue.

$PAR

END_TEXT

Context()->normalStrings;

Section::End();

###########################################

Section::Begin("Part 2:");

Context()->texStrings;
BEGIN_TEXT

Enter 2,3 below:
$BR

\{ $mans1->ans_rule(5) \},
\{ $mans1->ans_rule(5) \}

END_TEXT
Context()->normalStrings;

ANS( $mans1->cmp() );

Section::End();

Scaffold::End();

ENDDOCUMENT();
```

---

```
DOCUMENT();

loadMacros(
  "PGstandard.pl",
  "MathObjects.pl",
  "parserMultiAnswer.pl",
  "PGcourse.pl",
);

# Print problem number and point value (weight) for the problem
TEXT(beginproblem());

# Show which answers are correct and which ones are incorrect
$showPartialCorrectAnswers = 1;

Context("Numeric");

$a = 2;
$b = 3;

$ans1 = Real($a);
$ans2 = Real($b);

# ===============================================================


$mans1 = MultiAnswer( $ans1, $ans2 )->with(
  singleResult => 0,
  allowBlankAnswers => 1,
  checker => sub {
      my ( $correct, $student, $self, $ansHash ) = @_;
      my @stud_ans = @{$student};
      my @results = ( 0,0 );

      for ( $ii = 0; $ii < 2 ; $ii++ ) {
        if ( $$correct[ $ii ] == $$student[ $ii ] ) {
          $results[ $ii ] = 1;
        }
      }
      return( @results );
  }
);



# ===============================================================

Context()->texStrings;
BEGIN_TEXT

Text removed. There had been a question here. Not needed to trigger the issue.

$PAR

END_TEXT

Context()->normalStrings;


Context()->texStrings;
BEGIN_TEXT

Enter 2,3 below:
$BR

\{ $mans1->ans_rule(5) \},
\{ $mans1->ans_rule(5) \}

END_TEXT
Context()->normalStrings;

ANS( $mans1->cmp() );

ENDDOCUMENT();
```

**Comment by taniwallach** [comment]
@dpvc, @mgage, @drdrew42, @drgrice1 - Do any of you have an idea how to avoid this issue?

**Comment by taniwallach** [comment]
I just saw the same error message when an answer box was left blank in a scaffolded problem where `singleResult => 0`, `allowBlankAnswers => 0`, and `checkTypes => 0`. 

The specific problem has 8 input boxes in the scaffold section I was looking at (after a complaint), and any box being empty except the last triggered the warning. No warning was given when the last box was blank.

Testing the same answer after removing the scaffolding from the problem did not trigger the error message.

**Comment by drgrice1** [comment]
The issue is not with the lines that you showed in your comment.  The warning says line 380, but the issue is actually from line 378, and it is the fact that `$ans->{ans_message}` is compared as a string to the empty string when it is undefined.  I am not sure if you have noticed this, but the line numbers in warnings of this type are always one or two more than the actual line number that causes the warning.

I am not sure why the warning only occurs inside a scaffold though.  It is probably something to do with the way that scaffold evaluate the answers to determine if they are correct or not and hence if the scaffold should be open or not.

**Comment by taniwallach** [comment]
@drgrice1 - Thanks. The patch in https://github.com/openwebwork/pg/pull/625 fixes the issue.

>  I am not sure if you have noticed this, but the line numbers in warnings of this type are always one or two more than the actual line number that causes the warning.

I'm not sure if I noticed it in the past, but if so I had certainly forgotten about it. I'll try to remember in the future about this delta in the line number reported, as maybe then I would have figured out what to fix.

> I am not sure why the warning only occurs inside a scaffold though. It is probably something to do with the way that scaffold evaluate the answers to determine if they are correct or not and hence if the scaffold should be open or not.

It certainly must have something to do with differences in how scaffold processes things.

I looked and noticed that in `macros/scaffold.pl` inside `section_answers` the `Parser::Eval` call is being made only `if defined($input) && $input ne ""` so maybe when an input box of the MultiAnswer (except the last one?) is empty the `ans->{ans_message}` of the relevant answerHash was remaining undefined and then the code you fixed was later assuming it would be defined when that code was reached.

`section_answers` is called as part of the `Section::End` code. Adding some temporary warn calls before and after the call to `section_answers` and inside the routine confirms that the warning I reported above was issued while inside the `section_answers` call. In the test problem that was in the second section of the test problem, while processing `AnSwEr0002` which is the "end" of the MultiAnswer.
